### PR TITLE
Implement dynamic-k expert router

### PR DESCRIPTION
**Pull request by corey-shen**
NOTE: 
- This is assuming that the experts are in the last dimension
- Implementation is based on dynamic-k psuedo code/math proof
- Not sure if this works for dimensions > 1
- If dimension > 1 and we have a non-uniform distribution of experts being called, do 
we need to apply a masking or would each tensor maintain its fixed dimension?

TODO:
- Integrate with rest of the code (i.e. calling top_router() function)
- Testing

### CODE REVIEW

**Comment by ShonenMind** [comment]
Nice job with the implementation so far! Here's some feedback:

1) Line 244 will probably error, as it only works if our vector is one dimensional, any dimension higher and either there will be a run-time error or a random element might get picked (arguably worse because it's harder to debug)
2) p_sorted[:k] and idx_sorted[:k] will both splice using the same k, though it's not necessarily the case that every token has the same optimal_k value tho right?
Here's the code rewritten with some notes on it afterwards:

```
def top_router(self, logits, tau, max_k=4): 
        probabilities = torch.softmax(logits, dim=-1)
        p_sorted, idx_sorted = torch.sort(probabilities, descending=True, dim=-1)

        
        cumulative_sum = torch.cumsum(p_sorted, dim=-1)

        #we want to create a boolean mask to figure out the true "crossing" point!
        mask_keep = (cumsum < tau).type(torch.int32)
        mask_keep[..., 0] = 1 #this ellipsis notation represents all the leading dimensions btw 
        k_star = mask_keep.sum(dim = -1).clamp(max = max_k)


        #now, we'll need to start building up a matrix up to our max_k value in our function
        size_range = torch.arange(max_k, device=logits.device)
        select_the_mask = size_range[None, :].expand(*k_star.shape, max_k)
        #unsqueeze(-1) adds a size 1 dimension to the end of the tensor btw
        select_the_mask = select_the_mask < k_star.unsqueeze(-1)

        #at this point we pad our values as needed, specifically when select_the_mask is False
        pad_values = torch.full_like(idx_sort[..., :max_k], -1)
        select_idx = torch.where(select_the_mask, idx_sort[..., :max_k], pad_values)
        select_probs = torch.where(select_the_mask, p_sorted[..., max_k], pad_values, torch.zeros(1, dtype=logits.dtype, device=logits.device))


        renormalization = select_probs.sum(dim=-1, keepdim=True).clamp(min = 0.0000001)
        select_probs = select_probs / renormalization

        return select_idx, select_probs`
```

This reformatting of the code allows for all vectorization, so .item() is not needed. Also, select_the_mask now provides a mask PER token, allowing for the number of experts (up to the max k) to be different for every row. The outputs being padded in select_idx with 0 allows for those positions to essentially contribute nothing (so acting as if we're slicing them out like in the original code), however unlike slicing like how it was done previously, this new implementation still allows for information to propagate along these channels rather than being completely cut-off.

Also, the more I'm thinking about it, I think we actually CAN make tau not a hyperparameter constant that we choose, but instead it could be an actual learnable value. Consider adding something along the lines of `self.tau = nn.Parameter(torch.tensor(some_starting_value_here).logit())` and then `torch.sigmoid(self.tau)` during runtime.

**Comment by corey-shen** [comment]
I believe [torch.where](https://docs.pytorch.org/docs/stable/generated/torch.where.html) only takes in 3 arguments, but the following line has 4:
`select_probs = torch.where(select_the_mask, p_sorted[..., max_k], pad_values, torch.zeros(1, dtype=logits.dtype, device=logits.device))`

**Comment by corey-shen** [comment]
Code will be pushed/merged via a separate branch. Closing PR
